Order Editing:  Expand edit order UI test to include checking product and closing edit flow

thehenrybyrd opened the pull request:
## Description
Our existing UI test for Order Editing selects an existing order, taps the edit button, and checks the order title (order number). This PR adds another two steps to that test, to check for a product in the existing order, and check that "done" closes the editing flow.
Potential future work includes checks for editing items. This was deemed less urgent, as the flow is shared with Order Creation which already has testing around adding items.

Thanks @ealeksandrov working with me on this!

## Changes
- Add check for existing product.
- Add function to close the editing flow.
- Update UI test `test_edit_existing_order()` to check existing products and close editing flow.
- Renamed ` test_edit_existing_order()` to `test_edit_existing_order_loads_correct_data()` (as the test doesn't currently edit anything)

## Testing
Confirm test passes in CI.
Try breaking the flow, and make sure the test breaks. One way is to break the “done” button by commenting out `dismissHandler()` on line 172 of OrderForm. This will leave the edit screen up, and the test should fail.
Please let us know if you see anything that could be organized more clearly, or any places where it's unclear what we're testing.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ealeksandrov committed:
Add products check in Order Edit UI test

Co-Authored-By: Alicia Henry <15107387+thehenrybyrd@users.noreply.github.com>

ealeksandrov committed:
Add action to close Order Edit flow in UI test

Co-Authored-By: Alicia Henry <15107387+thehenrybyrd@users.noreply.github.com>

ealeksandrov committed:
Merge branch 'trunk' into update/edit-order-ui-test

Co-Authored-By: Alicia Henry <15107387+thehenrybyrd@users.noreply.github.com>

ealeksandrov committed:
Rename the test

Co-Authored-By: Alicia Henry <15107387+thehenrybyrd@users.noreply.github.com>

thehenrybyrd committed:
Update test coverage documentation for edit order UI test

Co-Authored-By: Evgeny Aleksandrov <3132438+ealeksandrov@users.noreply.github.com>

thehenrybyrd commented:
@jostnes @pachlava one thing I want to ask you about for this test is the final `.verifySingleOrderScreenLoaded()`.
Right now, this seems redundant since it just checks the `expectedElementGetters` (currently only `[ summaryCellGetter ]`) on `SingleOrderScreen`. We've just checked this, since `.closeEditingFlow()` returns `SingleOrderScreen`.

What's typical / best practice here? I see the rest of the tests in Woo Orders have some sort of "verifyScreenLoaded" check, but don't want to conform to that unless it's useful.

pachlava commented:
> I want to ask you about for this test is the final `.verifySingleOrderScreenLoaded()`.

Yes, as long as `expectedElementGetters` contains only one element, there's technically no difference between  `.closeEditingFlow()` returning `SingleOrderScreen` (which will wait only for the _first_ array element) and `verifyOrdersScreenLoaded()` which will eventually lead to waiting for _all_ array elements to appear.

However, expanding the `expectedElementGetters` with more elements will increase test coverage, but won't affect the `init` part, which still will wait only for the first element. 

One reason why I love having a line starting with `verify...` in the test body, is because it makes it obvious that this test verifies something. If it's removed, the test starts looking (on the surface) more like a sequence with no expected result, and it takes some time of digging into details to realize that there are many implicit assertions taking place anyway. (but this last paragraph only makes sense if the array is expanded).

If we compare the execution time, the time win/lose is around 1-2 seconds:
- in current state it's 31-32 seconds
- with `verifyOrdersScreenLoaded()` removed, it's 30-31
- with `expectedElementGetters` expanded with `editOrderButtonGetter` (not including `collectPaymentButtonGetter` because its presence depends on the order state) it's 32-33.

jostnes commented:
> What's typical / best practice here? I see the rest of the tests in Woo Orders have some sort of "verifyScreenLoaded" check, but don't want to conform to that unless it's useful.

On top of what @pachlava already mentioned above, I also think it's good practice to have assertion methods with an obvious name for readability. From the test, `.closeEditingFlow()` doesn't tell us what is being verified but `.verifySingleOrderScreenLoaded()` does immediately without going into the implementation.

jostnes commented on WooCommerce/UITestsFoundation/Screens/Orders/UnifiedOrderScreen.swift:
```diff
@@ -236,4 +236,26 @@ public final class UnifiedOrderScreen: ScreenObject {
+    /// Checks the screen for existence of all products, checking each name.
+    /// - Parameter productNames: Array of product names to check.
+    /// - Returns: Unified Order screen object.
+    @discardableResult
```
```suggestion
```

think we can remove the `@discardableResult` since this has a return value that we know and want to use.

jostnes commented on WooCommerce/UITestsFoundation/Screens/Orders/UnifiedOrderScreen.swift:
```diff
@@ -236,4 +236,26 @@ public final class UnifiedOrderScreen: ScreenObject {
+
+    /// Finishes Order Editing process
+    /// - Returns: Single Order Screen object.
+    @discardableResult
```
```suggestion
```

same as above, think this can be removed.

thehenrybyrd replied:
@jostnes as I was removing this and the previous `@discardableResult`, I realized that we have a lot of `@discardableResult`s throughout `UnifiedOrderScreen` and `SingleOrderScreen`.

My understanding is that  `@discardableResult` is used to suppress the warning when we have a `return` that we won't actually use. 
In our UI tests, much of the time we `return` a screen, and then the screen in question is displayed. So in those cases, our result is not discardable - we don't expect to get a warning that we aren't using the return, so we should not suppress such a warning.
In a few cases, we `return self`, in which case we do not change screens. In this case, we do not use the `return`, and we do expect to see a warning about this. Since we know why the warning appears, and we are certain this is the behavior we want, we can suppress the warning by using `@discardableResult`.

Am I understanding that correctly?

If so, I can go ahead and remove the unnecessary `@discardableResult`s from `UnifiedOrderScreen` and `SingleOrderScreen` - those for functions that do return a different screen.

jostnes replied:
Yup, that's correct - `@discardableResult` is used to suppress the warning when we have a return that we won't actually use. 

We return the screen value because as we are chaining the tests, each step in the chain is expecting a certain screen. Those `return` values are used to ensure that the test knows at that point, it's on which screen. This is for both cases when we return a different screen or `self` (which is equivalent to returning the screen it was originally on). In the cases where the method is the last step for the test, the chain stops, and the returned screen is not used anymore. In this case, we can use `@discardableResult` to suppress the warning (or we can even remove the return value since we know it's not going to be used and there are usually assertions in those methods)

So yup, you can go ahead and remove all the unnecessary `@discardableResult` from the screen. Thanks for cleaning it up!

ealeksandrov replied:
Adding my point of view: while removing `@discardableResult` won't trigger a warning in the current code, I think it serves additional documentation purpose. It means that annotated function is primarily used for its "side effect" and the return value is just a handy reference if you need it.

thehenrybyrd replied:
Thanks for the discussion! I've removed `@discardableResult` in cases where we return a different screen, and left it for methods that return `self`.

jostnes commented on WooCommerce/WooCommerceUITests/Tests/OrdersTests.swift:
```diff
@@ -42,13 +42,16 @@ final class OrdersTests: XCTestCase {
     }
 
-    func test_edit_existing_order() throws {
+    func test_edit_existing_order_loads_correct_data() throws {
```
```suggestion
    func test_load_existing_order() throws {
```

what do you think of updating test name to above? to match the naming convention of other tests where the verb is the second word in the test name.

thehenrybyrd replied:
Yep, much better name! Thanks for catching that.

jostnes reviewed (approved):
Thanks for expanding the test case @thehenrybyrd (and @ealeksandrov)! The changes look good and the test works well locally. Left some nitpicky comments, but other than that it's good to :shipit:

thehenrybyrd committed:
Update test name to match conventions

Co-Authored-By: Evgeny Aleksandrov <3132438+ealeksandrov@users.noreply.github.com>

thehenrybyrd committed:
Remove 2 `discardableResult`s from UnifiedOrderScreen

Co-Authored-By: Evgeny Aleksandrov <3132438+ealeksandrov@users.noreply.github.com>

thehenrybyrd committed:
Add additional ElementGetter to init

Co-Authored-By: Evgeny Aleksandrov <3132438+ealeksandrov@users.noreply.github.com>

thehenrybyrd commented:
Thanks @pachlava and @jostnes for your feedback. I totally agree with your reasoning about having a `verify...` line in the test. I've added an additional getter `editOrderButtonGetter` as suggested.

thehenrybyrd committed:
Remove unnecessary `discardableResult`s from SingleOrderScreen

`discardableResult` is unnecessary when we have a returned result which we expect to use.

Where we `return self`, the result `self` isn’t used directly, but staying on the screen allows us to chain methods. So `discardableResult` is used to suppress the warning.

Co-Authored-By: Evgeny Aleksandrov <3132438+ealeksandrov@users.noreply.github.com>

thehenrybyrd committed:
Remove unnecessary `discardableResult`s from UnifiedOrderScreen

`discardableResult` is unnecessary when we have a returned result which we expect to use.

Where we `return self`, the result `self` isn’t used directly, but staying on the screen allows us to chain methods. So `discardableResult` is used to suppress the warning.

Co-Authored-By: Evgeny Aleksandrov <3132438+ealeksandrov@users.noreply.github.com>

ealeksandrov committed:
Verify last navigation step in create_new_order test

Co-Authored-By: Alicia Henry <15107387+thehenrybyrd@users.noreply.github.com>
